### PR TITLE
update encfs readme

### PR DESCRIPTION
**Pull request by stevendongatmsft**

### CODE REVIEW

File: examples/encfs/aci-arm-template.json [hunk 1]
```diff
@@ -94,8 +94,7 @@
         "imageRegistryCredentials": [
           {
             "server": "<registry-name>",
-            "username": "",
```
Comment (hgarvison) [inline examples/encfs/aci-arm-template.json]: I wouldn't completely remove this since it's an option. Maybe just reference the options in the README?

File: examples/encfs/README.md [hunk 2]
```diff
@@ -55,7 +55,7 @@ Replace [managed-identity-with-right-permissions-to-key-vault](aci-arm-template.
 
 #### 3. Populate Image Registry Credentials
 
-Update the [image registry credentials](aci-arm-template.json?plain=1#L87) on the ARM template in order to access a private container registry. The credential could be either a managed identity or username/password. This section is not needed for public images. 
+Update the [image registry credentials](aci-arm-template.json?plain=1#L94) on the ARM template in order to access a private container registry. The credential could be either a managed identity or username/password. This section is not needed for public images. If the credential needs to be a managed identity, the managed identity needs to be granted access to the private images.
```
Comment (hgarvison) [inline examples/encfs/README.md]: I would show what this looks like in the arm template here

File: examples/encfs/README.md [hunk 3]
```diff
@@ -67,21 +67,32 @@ az account get-access-token --resource https://managedhsm.azure.net
...
+1. Generates a binary key file named `keyfile.bin` locally if such a file does not exist(see NOTE). 
+2. Creates `encfs.img` with the contents of the `generatefs/filesystem` directory.
+3. Encrypt the file system with `keyfile.bin` 
+
+NOTE: Users have the option of generating `keyfile.bin` on their own by issuing the following command: 
+
+```
+dd if=/dev/random of="<key-file-path>" count=1 bs=32
+```
+Or re-use a pre-existing `keyfile.bin`. In both cases, the script checks the existence of `keyfile.bin` and encrypts the `encfs.img` with it if such a file exists.
```
Comment (hgarvison) [inline examples/encfs/README.md]: I would mention that it checks for it in the generatefs directory

File: examples/encfs/README.md [hunk 4]
```diff
@@ -67,21 +67,32 @@ az account get-access-token --resource https://managedhsm.azure.net
...
 
 
 #### 6. Encrypted Filesystem
 The user needs to instantiate an [Azure Storage Container](https://learn.microsoft.com/en-us/azure/storage/common/storage-account-create?toc=%2Fazure%2Fstorage%2Fblobs%2Ftoc.json&bc=%2Fazure%2Fstorage%2Fblobs%2Fbreadcrumb%2Ftoc.json&tabs=azure-portal) onto which the encrypted filesystem will be uploaded. The roles *Reader* and *Storage Blob Reader* roles need to be assigned to the user-assigned managed identity. Additionally, the role of *Storage Blob Contributor* needs to be assigned for a read-write filesystem.
 
-The script `generatefs/generatefs.sh` creates `encfs.img` with the contents of the `generatefs/filesystem` directory. You may need to adjust the size of the image in the script, as it isn't calculated automatically. 
+The script `generatefs/generatefs.sh` does the following in order: 
 
-The script expects a symmetric key stored in binary format `keyfile.bin` previously created during key import phase. If not passed, the script will generate a new one and the user will need to follow the import key instructions on the new key.
+1. Generates a binary key file named `keyfile.bin` locally if such a file does not exist(see NOTE).
```
Comment (hgarvison) [inline examples/encfs/README.md]: put a space after 'exist'

File: examples/encfs/README.md [hunk 5]
```diff
@@ -67,21 +67,32 @@ az account get-access-token --resource https://managedhsm.azure.net
...
 
-The script `generatefs/generatefs.sh` creates `encfs.img` with the contents of the `generatefs/filesystem` directory. You may need to adjust the size of the image in the script, as it isn't calculated automatically. 
+The script `generatefs/generatefs.sh` does the following in order: 
 
-The script expects a symmetric key stored in binary format `keyfile.bin` previously created during key import phase. If not passed, the script will generate a new one and the user will need to follow the import key instructions on the new key.
+1. Generates a binary key file named `keyfile.bin` locally if such a file does not exist(see NOTE). 
+2. Creates `encfs.img` with the contents of the `generatefs/filesystem` directory.
```
Comment (hgarvison) [inline examples/encfs/README.md]: State that encfs.img is the encrypted filesystem image

File: examples/encfs/README.md [hunk 6]
```diff
@@ -67,21 +67,32 @@ az account get-access-token --resource https://managedhsm.azure.net
...
+3. Encrypt the file system with `keyfile.bin` 
+
+NOTE: Users have the option of generating `keyfile.bin` on their own by issuing the following command: 
+
+```
+dd if=/dev/random of="<key-file-path>" count=1 bs=32
+```
+Or re-use a pre-existing `keyfile.bin`. In both cases, the script checks the existence of `keyfile.bin` and encrypts the `encfs.img` with it if such a file exists. 
+
+You may need to adjust the size of the image in the script, as it isn't calculated automatically.
```
Comment (hgarvison) [inline examples/encfs/README.md]: Link it to the line in the script: https://github.com/microsoft/confidential-sidecar-containers/blob/2c76c71ab6610dbabb601ac0c1c9986df1e84151/examples/encfs/generatefs/generatefs.sh#L26

File: examples/encfs/README.md [hunk 7]
```diff
@@ -122,7 +133,7 @@ At this point, the `encfs-sidecar-args.json` file should be completely filled ou
...
+Configure the [LogFile](aci-arm-template.json#L39) and [LogLevel](aci-arm-template.json#L43) fields on the `aci-arm-template.json`. eg. `log.txt` as the value of [LogFile] so that users can cat `log.txt` at the root of encrypted-filesystem-sidecar-container container. At this point, the `aci-arm-template.json` file should be filled out except for the `ccepolicy` field. Set the value of the `ccePolicy` to an empty string. After installing the [Azure `confcom` CLI extension](#policy-generation), run the following command to generate the security policy and include the `--debug-mode` option so that the policy allows users to shell into the container. The --debug-mode option is only needed for testing purposes and should not be used in production, as it allows users to shell into the containers.
```
Comment (hgarvison) [inline examples/encfs/README.md]: Note that the defaults are to print to stderr and at the warning level.

Comment (stevendongatmsft) [reply examples/encfs/README.md]: How to use the default? Currently on our template, it is listed as the following: 
{
                  "name": "LogLevel",
                  "value": "<optional-loglevel-trace-debug-info-warning-error-fatal-panic>"
}
Would it use default if users leave it as an empty string?

File: examples/encfs/README.md (hunk 2 above)
Comment (stevendongatmsft) [reply examples/encfs/README.md]: Yeah the arm template is also updated with managed identity if that's what you mean

Comment (hgarvison) [reply examples/encfs/README.md]: No, what I mean is to show how you can use the managed identity OR the username/password here, since you removed username/password from the arm template.

File: examples/encfs/README.md (hunk 7 above)
Comment (hgarvison) [reply examples/encfs/README.md]: Yes, the default is used if it's an empty string https://github.com/microsoft/confidential-sidecar-containers/blob/2c76c71ab6610dbabb601ac0c1c9986df1e84151/cmd/skr/main.go#L41.

File: examples/encfs/README.md [hunk 8]
```diff
@@ -7,16 +7,18 @@
...
 ```
 az confcom acipolicygen -a aci-skr-arm-template.json --debug-mode
 ```
 
-The ARM template file file includes two entries: (i) encrypted filesystem sidecar container which whitelists the /encfs.sh as entry point command and the environment variable *EncfsSideCarArgs* used by the script, and (ii) an application container which whitelists a while loop command as entry point command.
+The ARM template file includes two entries: (i) encrypted filesystem sidecar container which whitelists the /encfs.sh as entry point command and the environment variable *EncfsSideCarArgs* used by the script, and (ii) an application container which whitelists a while loop command as entry point command.
 
 ### Step by Step Example 
 
+This example is made to be run on Linux/WSL. Please also make sure the the scripts used in this example are set to LF instead of CRLF.
```
Comment (hgarvison) [inline examples/encfs/README.md]: I would say: "Also make sure the line endings in the scripts are set to LF instead of CRLF" for the second sentence.

File: examples/encfs/README.md [hunk 9]
```diff
@@ -66,22 +68,34 @@ The AAD token with permission to AKV/mHSM can be obtained with the following com
 az account get-access-token --resource https://managedhsm.azure.net
 ```
 
-Replace [AAD token](importkeyconfig.json#L11) in `importkeyconfig.json` with the output accessToken.
-Replace [ENCFS AKV endpoint](encfs-sidecar-args.json#L15) in `enfcs-sidecar-args.json` with the mhsm endpoint.
+Replace [AAD token](importkeyconfig.json#L11) and [ENCFS AKV endpoint](importkeyconfig.json#L9) in `importkeyconfig.json` with the output accessToken.
```
Comment (SethHollandsworth) [inline examples/encfs/README.md]: Should [ENCFS AKV endpoint](importkeyconfig.json#L9) also be replaced with the output accessToken? that's what it reads like

Comment (SethHollandsworth) [reply examples/encfs/README.md]: also this chunk seems like it would fit better in step 5

File: examples/encfs/README.md [hunk 10]
```diff
@@ -66,22 +68,34 @@ The AAD token with permission to AKV/mHSM can be obtained with the following com
...
 
 
 #### 6. Encrypted Filesystem
 The user needs to instantiate an [Azure Storage Container](https://learn.microsoft.com/en-us/azure/storage/common/storage-account-create?toc=%2Fazure%2Fstorage%2Fblobs%2Ftoc.json&bc=%2Fazure%2Fstorage%2Fblobs%2Fbreadcrumb%2Ftoc.json&tabs=azure-portal) onto which the encrypted filesystem will be uploaded. The roles *Reader* and *Storage Blob Reader* roles need to be assigned to the user-assigned managed identity. Additionally, the role of *Storage Blob Contributor* needs to be assigned for a read-write filesystem.
 
-The script `generatefs/generatefs.sh` creates `encfs.img` with the contents of the `generatefs/filesystem` directory. You may need to adjust the size of the image in the script, as it isn't calculated automatically. 
+The script `generatefs/generatefs.sh` does the following in order:
```
Comment (SethHollandsworth) [inline examples/encfs/README.md]: I think is where we should have something about needing to run this script on WSL/Linux

File: examples/encfs/README.md [hunk 11]
```diff
@@ -122,13 +136,13 @@ At this point, the `encfs-sidecar-args.json` file should be completely filled ou
...
+Configure the [LogFile](aci-arm-template.json#L39) and [LogLevel](aci-arm-template.json#L43) fields on the `aci-arm-template.json`. eg. `log.txt` as the value of [LogFile] so that users can cat `log.txt` at the root of encrypted-filesystem-sidecar-container container. The default logging is `warning` level. At this point, the `aci-arm-template.json` file should be filled out except for the `ccepolicy` field. Set the value of the `ccePolicy` to an empty string. After installing the [Azure `confcom` CLI extension](#policy-generation), run the following command to generate the security policy and include the `--debug-mode` option so that the policy allows users to shell into the container. The --debug-mode option is only needed for testing purposes and should not be used in production, as it allows users to shell into the containers.
```
Comment (SethHollandsworth) [inline examples/encfs/README.md]: different casing between `ccepolicy` and `ccePolicy` in this line

Comment (SethHollandsworth) [reply examples/encfs/README.md]: the second usage of `--debug-mode` should have backticks. Also I would rephrase to something like, "If necessary when debugging, users may add the `--debug-mode` flag to allow shelling into the containers. Note this should only be used for debugging and not recommended for production systems".

That way it doesn't sound like the default thing to do

Comment (ksayid) [reply examples/encfs/README.md]: should we also add this caveat to the policy generation section (before the step-by-step example)? I'm looking at line 10.

File: examples/encfs/README.md [hunk 12]
```diff
@@ -7,16 +7,18 @@
...
+The ARM template can be used directly to generate a security policy. The following command generates a security policy and automatically injects it into the template. Make sure `--debug-mode` option is included so that the generated policy allows shelling into container to see the released key in this example.
 
 ```
 az confcom acipolicygen -a aci-skr-arm-template.json --debug-mode
 ```
 
-The ARM template file file includes two entries: (i) encrypted filesystem sidecar container which whitelists the /encfs.sh as entry point command and the environment variable *EncfsSideCarArgs* used by the script, and (ii) an application container which whitelists a while loop command as entry point command.
+The ARM template file includes two entries: (i) encrypted filesystem sidecar container which whitelists the /encfs.sh as entry point command and the environment variable *EncfsSideCarArgs* used by the script, and (ii) an application container which whitelists a while loop command as entry point command.
```
Comment (ksayid) [inline examples/encfs/README.md]: nit:  /encfs.sh should have backticks

File: examples/encfs/README.md [hunk 13]
```diff
@@ -7,16 +7,18 @@
...
 
 ### Step by Step Example 
 
+This example is made to be run on Linux/WSL. Also make sure the line endings in the scripts used in this example are set to LF instead of CRLF. 
+
 Here is an example of running the encfs sidecar on confidential ACI. The MAA endpoint is the value of env var [Authority Endpoint](encfs-sidecar-args.json#L12). 
 The managed HSM instance endpoint corresponds to [AKV Endpoint](encfs-sidecar-args.json#L15). We will also define and import a key into the managed HSM as shown in [`importkeyconfig.json`](importkeyconfig.json#L2)
```
Comment (ksayid) [inline examples/encfs/README.md]: nit: managed HSM is mentioned here but defined later (line 39)

File: examples/encfs/README.md [hunk 14]
```diff
@@ -122,13 +136,13 @@ At this point, the `encfs-sidecar-args.json` file should be completely filled ou
...
 
 ```
 az confcom acipolicygen -a aci-arm-template.json --debug-mode
 ```
 
-This should  automatically populate the [cce policy](aci-arm-template.json#L106) field of `aci-arm-template.json.` If you run the tool on an arm template with the cce policy field already populated, the tool will prompt you to overwrite the existing policy.
+This should automatically populate the [cce policy](aci-arm-template.json#L106) field of `aci-arm-template.json.` If you run the tool on an arm template with the cce policy field already populated, the tool will prompt you to overwrite the existing policy.
```
Comment (ksayid) [inline examples/encfs/README.md]: nit: cce policy should be `ccePolicy` to be consistent with line 139 and the aci arm template
nit: arm template should be ARM template to be consistent with earlier naming

File: examples/encfs/README.md [hunk 15]
```diff
@@ -176,3 +189,23 @@ Alternatively, the whitelisted command in test-encfs-container outputs the follo
 This is a file inside the filesystem.
 This is a file inside the filesystem.
 ```
+
+### Running the ENCFS example with RSA Key with Salt
```
Comment (hgarvison) [inline examples/encfs/README.md]: This should just be in step 5 instead of at the bottom.

File: examples/encfs/README.md (hunk 10 above)
Comment (stevendongatmsft) [reply examples/encfs/README.md]: Yeah this is mentioned in the beginning of the example

File: examples/encfs/README.md (hunk 9 above)
Comment (stevendongatmsft) [reply examples/encfs/README.md]: good catch

File: examples/encfs/README.md (hunk 15 above)
Comment (stevendongatmsft) [reply examples/encfs/README.md]: I put this here because RSA does not follow the regular oct sequence of operation. So it will be best for users to look at this section at the end.

Comment (hgarvison) [reply examples/encfs/README.md]: Putting RSA here, means you're putting key generation steps out of order and a user who wants to do it this way will have to scroll to the bottom to find this info.

Comment (hgarvison) [reply examples/encfs/README.md]: Also the steps are almost identical with the exception of adding the salt and specifying that it's and RSA key

File: examples/encfs/README.md [hunk 16]
```diff
@@ -66,22 +75,88 @@ The AAD token with permission to AKV/mHSM can be obtained with the following com
...
+```
+go run tools/importkey/main.go -c importkeyconfig.json -out=true
+```
+If user configured `RSA-HSM` key type on `importkeyconfig.json`, make sure the number of characters of the salt string is even number. 
+Upon successful import completion, you should see something similar to the following: 
+
+```
+[34 71 33 117 113 25 191 84 199 236 137 166 201 103 83 20 203 233 66 236 121 110 223 2 122 99 106 20 22 212 49 224]
+https://<mhsm-name>.managedhsm.azure.net/keys/doc-sample-key-release/8659****0cdff08
+{"version":"0.2","anyOf":[{"authority":"https://sharedeus2.eus2.test.attest.azure.net","allOf":[{"claim":"x-ms-sevsnpvm-hostdata","equals":"aaa7***7cc09d"},{"claim":"x-ms-compliance-status","equals":"azure-compliant-uvm"},{"claim":"x-ms-sevsnpvm-is-debuggable","equals":"false"}]}]}
```
Comment (hgarvison) [inline examples/encfs/README.md]: Update the authority part here to remove the "sharedeus2" name.

File: examples/encfs/README.md [hunk 17]
```diff
@@ -66,22 +75,88 @@ The AAD token with permission to AKV/mHSM can be obtained with the following com
...
+```
+
+In this case, use the following commands to verify the key has been successfully imported: 
+
+```
+az account set --subscription "<SUBSCRIPTION>"
+az keyvault key list --hsm-name <mHSM NAME> -o table
+```
+The main.go golang script generates a private rsa key or an oct key named `keyfile.bin` based on the key type configuration on `importkeyconfig.json`.
```
Comment (hgarvison) [inline examples/encfs/README.md:146]: Until we update the importkey tool, it doesn't create the oct key. generatefs.sh creates the oct key. So just remove the mention of oct key for now and once we merge this, I can work to make it so that importkey creates the keys and I'll update the README accordingly.

Comment (stevendongatmsft) [reply examples/encfs/README.md:146]: The import key tool actually also create the oct key as well.

File: examples/encfs/README.md (hunk 16 above)
Comment (stevendongatmsft) [reply examples/encfs/README.md]: What should we put authority here then?

Comment (hgarvison) [reply examples/encfs/README.md]: Just put "authority name."

File: examples/encfs/README.md (hunk 17 above)
Comment (hgarvison) [reply examples/encfs/README.md:146]: Huh, then we should just remove key generation from the generatefs.sh script

**Comment by stevendongatmsft** [comment]
@hgarvison ok, the 2 issues and merge conflicts are resolved

File: examples/encfs/generatefs/generatefs.sh [hunk 18]
```diff
@@ -12,8 +12,8 @@ deviceNamePath="/dev/mapper/$deviceName"
 if [ -f "$keyFilePath" ]; then
     echo "keyfile exists"
 else
-    echo "[!] Generating keyfile..."
-    dd if=/dev/random of="$keyFilePath" count=1 bs=32
+    echo "[!] keyfile does not exist."
+    exist 1
```
Comment (hgarvison) [inline examples/encfs/generatefs/generatefs.sh]: exit?

File: examples/encfs/README.md [hunk 19]
```diff
@@ -66,22 +75,88 @@ The AAD token with permission to AKV/mHSM can be obtained with the following com
...
+Replace [ENCFS AKV endpoint](encfs-sidecar-args.json#L15) in `enfcs-sidecar-args.json` with the mHSM endpoint.
+
+After setting up an Azure Key Vault resource, fill in the `importkeyconfig.json` file with the name of the key to be created and imported into the key vault [Key name](importkeyconfig.json#L3). 
+Additionally, fill in `encfs-sidecar-args.json` with the name of the key to be created and imported into the key vault [Key name](encfs-sidecar-args.json#L9).
+
+Additionally, fill in the optional [key derivation](importkeyconfig.json#L14) for RSA keys and [Key type: `RSA-HSM` or `oct-HSM`](importkeyconfig.json#L4) fields or remove these fields from the `importkeyconfig.json` file. 
+Fill in [key derivation](encfs-sidecar-args.json#L18) and [Key type: `RSA-HSM` or `oct-HSM`](encfs-sidecar-args.json#L10) in `encfs-sidecar-args.json` or remove them as well as well. 
+
+The user needs to instantiate an [Azure Storage Container](https://learn.microsoft.com/en-us/azure/storage/common/storage-account-create?toc=%2Fazure%2Fstorage%2Fblobs%2Ftoc.json&bc=%2Fazure%2Fstorage%2Fblobs%2Fbreadcrumb%2Ftoc.json&tabs=azure-portal) onto which the encrypted filesystem will be uploaded.
```
Comment (hgarvison) [inline examples/encfs/README.md:91]: Since this isn't part of creating the key, maybe we can put this as a step before this: "5. Create Azure Storage Container"

File: examples/encfs/README.md [hunk 20]
```diff
@@ -66,22 +75,88 @@ The AAD token with permission to AKV/mHSM can be obtained with the following com
...
+```
+go run tools/importkey/main.go -c importkeyconfig.json -out=true
+```
+If user configured `RSA-HSM` key type on `importkeyconfig.json`, make sure the number of characters of the salt string is even number. 
+Upon successful import completion, you should see something similar to the following: 
+
+```
+[34 71 33 117 113 25 191 84 199 236 137 166 201 103 83 20 203 233 66 236 121 110 223 2 122 99 106 20 22 212 49 224]
+https://<mhsm-name>.managedhsm.azure.net/keys/doc-sample-key-release/8659****0cdff08
+{"version":"0.2","anyOf":[{"authority":"<authority-name>","allOf":[{"claim":"x-ms-sevsnpvm-hostdata","equals":"aaa7***7cc09d"},{"claim":"x-ms-compliance-status","equals":"azure-compliant-uvm"},{"claim":"x-ms-sevsnpvm-is-debuggable","equals":"false"}]}]}
```
Comment (hgarvison) [inline examples/encfs/README.md]: Oops, I meant <authority-name>.rest_of_url, similar to what you did with the mhsm, but this is prob okay or just use whatever we have for the field in the import key config file.

File: examples/encfs/README.md (hunk 19 above)
Comment (stevendongatmsft) [reply examples/encfs/README.md:91]: You're talking about separating step 5 into 2 sub steps. One is for fill in the key information and the other is create azure storage container?

File: examples/encfs/README.md [hunk 21]
```diff
@@ -66,22 +75,88 @@ The AAD token with permission to AKV/mHSM can be obtained with the following com
...
+
+```
+az account set --subscription "<SUBSCRIPTION>"
+az keyvault key list --hsm-name <mHSM NAME> -o table
+```
+The main.go golang script generates a private rsa key or an oct key named `keyfile.bin` based on the key type configuration on `importkeyconfig.json`.
+It then uploads the binary key file named `keyfile.bin` to the mHSM under the [key ID](importkeyconfig.json#L3) along with the key released policy from `importkeyconfig.json`. 
 
-Additionally, fill in the optional [key derivation](importkeyconfig.json#L14) for RSA keys and [Key type: `RSA-HSM` or `oct-HSM`](importkeyconfig.json#L4) fields or remove these fields from the `importkeyconfig.json` file. Fill in [key derivation](encfs-sidecar-args.json#L18) and [Key type: `RSA-HSM` or `oct-HSM`](encfs-sidecar-args.json#L10) in `encfs-sidecar-args.json` as well.
+Copy the `keyfile.bin` to `generatefs/generatefs.sh` folder:
```
Comment (hgarvison) [inline examples/encfs/README.md]: I would just put this as part of step 8

File: examples/encfs/README.md (hunk 19 above)
Comment (hgarvison) [reply examples/encfs/README.md:91]: yes, it'll make smaller easier to follow steps

File: examples/encfs/README.md (hunk 20 above)
Comment (stevendongatmsft) [reply examples/encfs/README.md]: I will change it to <authority-url-name> because the urls for authority is pretty different from one to another.

File: examples/encfs/README.md (hunk 19 above)
Comment (stevendongatmsft) [reply examples/encfs/README.md:91]: ok

**Review by hgarvison — APPROVED** [review]
